tsl/src/remote/connection.c

```diff
@@ -1468,6 +1486,90 @@ add_userinfo_to_server_options(ForeignServer *server, Oid user_id)
 	return options;
 }
 
+/*
+ * Append the given string to the buffer, with suitable quoting for passing
+ * the string as a value in a keyword/value pair in a libpq connection string.
+ *
+ * The implementation is based on libpq appendConnStrVal().
+ */
+static void
+remote_connection_append_connstr_value(StringInfo buf, const char *str)
+{
+	const char *s;
+	bool needquotes;
+
+	/*
+	 * If the string is one or more plain ASCII characters, no need to quote
+	 * it. This is quite conservative, but better safe than sorry.
+	 */
+	needquotes = true;
+	for (s = str; *s; s++)
+	{
+		if (!((*s >= 'a' && *s <= 'z') || (*s >= 'A' && *s <= 'Z') || (*s >= '0' && *s <= '9') ||
+			  *s == '_' || *s == '.'))
+		{
+			needquotes = true;
+			break;
+		}
+		needquotes = false;
+	}
+
+	if (needquotes)
+	{
+		appendStringInfoChar(buf, '\'');
+		while (*str)
+		{
+			/* ' and \ must be escaped by to \' and \\ */
+			if (*str == '\'' || *str == '\\')
+				appendStringInfoChar(buf, '\\');
+
+			appendStringInfoChar(buf, *str);
+			str++;
+		}
+		appendStringInfoChar(buf, '\'');
+	}
+	else
+		appendStringInfoString(buf, str);
+}
+
+char *
+remote_connection_get_connstr(const char *node_name)
+{
+	ForeignServer *server;
+	List *connection_options;
+	const char **keywords;
+	const char **values;
+	StringInfoData connstr;
+	StringInfoData connstr_escape;
+	int i;
+
+	server = data_node_get_foreign_server(node_name, ACL_NO_CHECK, false, false);
+	connection_options = add_userinfo_to_server_options(server, GetUserId());
+	setup_full_connection_options(connection_options, &keywords, &values);
+
+	/* Cycle through the options and create the connection string */
+	initStringInfo(&connstr);
+	i = 0;
+	while (keywords[i] != NULL)
+	{
+		appendStringInfo(&connstr, " %s=", keywords[i]);
+		remote_connection_append_connstr_value(&connstr, values[i]);
+		i++;
+	}
+	Assert(keywords[i] == NULL && values[i] == NULL);
+
+	initStringInfo(&connstr_escape);
+	enlargeStringInfo(&connstr_escape, connstr.len * 2 + 1);
+	connstr_escape.len += PQescapeString(connstr_escape.data, connstr.data, connstr.len);
```

Comment:
i would prefer to see the sql escaping in the caller where you build the sql string, but other than that looks good